File: version_java/src/main/java/com/example/version_java/ui/login/LoginActivity.java
package com.example.version_java.ui.login;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;

import com.example.version_java.R;
import com.example.version_java.base.BaseActivity;
import com.example.version_java.databinding.ActivityLoginBinding;
import com.example.version_java.ui.home.HomeActivity;
import com.example.version_java.ui.register.RegisterActivity;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class LoginActivity extends BaseActivity {

    private LoginViewModel loginViewModel;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initUi();
        initViewModel();
    }

    private void initUi() {
        ((ActivityLoginBinding) getBinding()).inputPassLogin.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                loginViewModel.login();
                return true;
            } else {
                return false;
            }
        });
    }

    private void initViewModel() {
        loginViewModel = new ViewModelProvider(this).get(LoginViewModel.class);
        ((ActivityLoginBinding) getBinding()).setViewModel(loginViewModel);

        loginViewModel.getViewStateLiveData().observe(LoginActivity.this, viewState -> {
            if (viewState instanceof LoginViewState) {
                onChangedViewState((LoginViewState) viewState);
            }
        });
    }

    private void onChangedViewState(LoginViewState viewState) {
        if (viewState instanceof LoginViewState.RouteRegister) {
            Intent intent = new Intent(LoginActivity.this, RegisterActivity.class);
            startActivity(intent);
        } else if (viewState instanceof LoginViewState.RouteHome) {
            Intent intent = new Intent(LoginActivity.this, HomeActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK);
            startActivity(intent);
        } else if (viewState instanceof LoginViewState.Error) {
            Toast.makeText(this, ((LoginViewState.Error) viewState).getMessage(), Toast.LENGTH_SHORT).show();
        } else if (viewState instanceof LoginViewState.EnableInput) {
            ((ActivityLoginBinding) getBinding()).inputEmailLogin.setEnabled(((LoginViewState.EnableInput) viewState).isEnable());
            ((ActivityLoginBinding) getBinding()).inputPassLogin.setEnabled(((LoginViewState.EnableInput) viewState).isEnable());
        } else if (viewState instanceof LoginViewState.ShowProgress) {
            ((ActivityLoginBinding) getBinding()).progressbar.bringToFront();
            ((ActivityLoginBinding) getBinding()).progressbar.setVisibility(View.VISIBLE);
        } else if (viewState instanceof LoginViewState.HideProgress) {
            ((ActivityLoginBinding) getBinding()).progressbar.setVisibility(View.GONE);
        }
    }


    public LoginActivity() {
        super(R.layout.activity_login);
        this.setBinding((ActivityLoginBinding) this.getBinding());
    }
}

File: app/src/main/java/com/example/crimenotification/ui/login/LoginActivity.kt
package com.example.crimenotification.ui.login

import android.content.Intent
import android.os.Bundle
import android.view.inputmethod.EditorInfo
import androidx.activity.viewModels
import androidx.core.view.isVisible
import com.example.crimenotification.R
import com.example.crimenotification.base.BaseActivity
import com.example.crimenotification.databinding.ActivityLoginBinding
import com.example.crimenotification.ext.showToast
import com.example.crimenotification.ui.home.HomeActivity
import com.example.crimenotification.ui.register.RegisterActivity
import com.example.crimenotification.ui.register.RegisterViewState
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class LoginActivity : BaseActivity<ActivityLoginBinding>(R.layout.activity_login) {

    private val loginViewModel by viewModels<LoginViewModel>()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        initUi()
        initViewModel()
    }

    private fun initUi() {
        binding.inputPassLogin.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                loginViewModel.login()
                true
            } else {
                false
            }
        }
    }

    private fun initViewModel() {
        binding.viewModel = loginViewModel
        loginViewModel.viewStateLiveData.observe(this) { viewState ->
            (viewState as? LoginViewState)?.let {
                onChangedLoginViewState(
                    viewState
                )
            }
        }
    }

    private fun onChangedLoginViewState(viewState: LoginViewState) {
        when (viewState) {

            is LoginViewState.Error -> {
                showToast(message = viewState.message)
            }

            is LoginViewState.EnableInput -> {
                with(binding) {
                    inputEmailLogin.isEnabled = viewState.isEnable
                    inputPassLogin.isEnabled = viewState.isEnable
                }
            }

            is LoginViewState.RouteHome -> {
                startActivity(Intent(this@LoginActivity, HomeActivity::class.java).apply {
                    addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
                })
            }

            is LoginViewState.RouteRegister -> {
                with(binding) {
                    inputEmailLogin.text.clear()
                    inputPassLogin.text.clear()
                }
                startActivity(Intent(this@LoginActivity, RegisterActivity::class.java))
            }

            is LoginViewState.ShowProgress -> {
                binding.progressbar.bringToFront()
                binding.progressbar.isVisible = true
            }

            is LoginViewState.HideProgress -> {
                binding.progressbar.isVisible = false
            }
        }
    }
}